File: api/app/Models/Deploy.php
<?php

declare(strict_types=1);

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Deploy extends Model
{
    use HasFactory, HasUuids;

    protected $fillable = [
        'site_id',
        'status',
        'git_branch',
        'commit_hash',
        'started_at',
        'completed_at',
        'logs',
        'error',
        'metadata',
    ];

    protected $casts = [
        'started_at' => 'datetime',
        'completed_at' => 'datetime',
        'logs' => 'array',
        'metadata' => 'array',
    ];

    public const STATUS_PENDING = 'pending';
    public const STATUS_IN_PROGRESS = 'in_progress';
    public const STATUS_SUCCESS = 'success';
    public const STATUS_FAILED = 'failed';
    public const STATUS_ROLLED_BACK = 'rolled_back';

    public const STATUSES = [
        self::STATUS_PENDING,
        self::STATUS_IN_PROGRESS,
        self::STATUS_SUCCESS,
        self::STATUS_FAILED,
        self::STATUS_ROLLED_BACK,
    ];

    public function site()
    {
        return $this->belongsTo(Site::class);
    }

    public function addLog(string $message): void
    {
        $logs = $this->logs ?? [];
        $logs[] = [
            'timestamp' => now()->toIso8601String(),
            'message' => $message,
        ];
        $this->logs = $logs;
        $this->save();
    }

    public function markAsInProgress(): void
    {
        $this->update([
            'status' => self::STATUS_IN_PROGRESS,
            'started_at' => now(),
        ]);
    }

    public function markAsSuccess(): void
    {
        $this->update([
            'status' => self::STATUS_SUCCESS,
            'completed_at' => now(),
        ]);
    }

    public function markAsFailed(string $error): void
    {
        $this->update([
            'status' => self::STATUS_FAILED,
            'completed_at' => now(),
            'error' => $error,
        ]);
    }
}

File: api/models/deploy.py
"""
Deploy Modelleri
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class DeployStatus(str, Enum):
    """Deploy durumları"""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    SUCCESS = "success"
    FAILED = "failed"
    ROLLED_BACK = "rolled_back"


class DeployRequest(BaseModel):
    """Deploy isteği modeli"""
    site_id: str = Field(..., description="Site ID")
    git_branch: Optional[str] = Field(None, description="Deploy edilecek branch (varsayılan: site branch)")
    force: bool = Field(False, description="Zorla deploy et")
    run_migrations: bool = Field(False, description="Database migration'ları çalıştır (Laravel için)")
    clear_cache: bool = Field(True, description="Cache'leri temizle")
    install_dependencies: bool = Field(True, description="Bağımlılıkları yükle")
    
    class Config:
        json_schema_extra = {
            "example": {
                "site_id": "example-com",
                "git_branch": "main",
                "force": False,
                "run_migrations": True,
                "clear_cache": True,
                "install_dependencies": True
            }
        }


class DeployResponse(BaseModel):
    """Deploy yanıt modeli"""
    deploy_id: str
    site_id: str
    status: DeployStatus
    message: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    logs: List[str] = []
    error: Optional[str] = None
    
    class Config:
        use_enum_values = True


class DeployLog(BaseModel):
    """Deploy log modeli"""
    timestamp: datetime
    level: str
    message: str
    data: Optional[Dict[str, Any]] = None

